### PR TITLE
fix for bug:2042

### PR DESCRIPTION
**Pull request by sunilk747**
License analysis is getting failed because of incorrect splitting while adding node to graph. This fix takes care of the splitting of licenses correctly.
```
If license is "Apache License, Version 2"
Result : ["Apache License", "Version 2"]   (incorrect)
It should be 
["Apache License, Version 2"]   (Correct)
```

### CODE REVIEW

File: src/graph_populator.py [hunk 1]
```diff
@@ -106,7 +106,16 @@ def construct_version_query(cls, input_json):
                     declared_licenses = details[0]['declared_licenses']
                 elif details[0].get('declared_license'):
                     # string with comma separated license names
-                    declared_licenses = details[0]['declared_license'].split(',')
+                    # declared_licenses = details[0]['declared_license'].split(',')
+                    declared_licenses = details[0].get('declared_license').split(', ')
+                    final_declared_licenses = list()
+                    for i in xrange(0, len(declared_licenses)):
+                        if declared_licenses[i].startswith("Version"):
```
Comment (sara-02) [inline src/graph_populator.py]: Starts with `Version` or `version`, would be better to consider both cases.

File: src/graph_populator.py [hunk 2]
```diff
@@ -106,7 +106,16 @@ def construct_version_query(cls, input_json):
...
+                            final_declared_licenses[-1] = final_declared_licenses[-1] + ", "\
```
Comment (sara-02) [inline src/graph_populator.py]: What if we have something like `Apache, Version2 , MIT`. Here we want declared list to be `Apache Version 2, MIT`.
Maybe fix it by `final_delcared_licenses[i-1]`=final_declared_licenses[i-1]+declared_licenses[i]

File: src/graph_populator.py [hunk 3]
```diff
@@ -106,7 +106,16 @@ def construct_version_query(cls, input_json):
                     declared_licenses = details[0]['declared_licenses']
                 elif details[0].get('declared_license'):
```
Comment (sara-02) [inline src/graph_populator.py]: @miteshvp  @sunilk747  How is the `elseif` condition different from the `if` condition above :/ 
https://github.com/sunilk747/fabric8-analytics-data-model/blob/49fd9a6ff7509f16e93042624b1dc5ec6fba3b0b/src/graph_populator.py#L104

File: src/graph_populator.py (hunk 1 above)
Comment (sunilk747) [reply src/graph_populator.py]: This case, I am planning to handle in the license analysis repo in another PR.

**Comment by sara-02** [comment]
@sunilk747 thanks for clearing the doubts. LGTM. @sunilk747 the `if-elif` condition still remains a concern though.
@miteshvp please take a look.

File: src/graph_populator.py (hunk 1 above)
Comment (jpopelka) [inline src/graph_populator.py]: If you use the `i` only as index into the `declared_licenses`, then you access the members directly via

```python
for declared_license in declared_licenses:
    if declared_license.startswith("Version"):
```

Or if you really need the `i` also for some other reason(s), then you can use [enumerate](https://docs.python.org/3/library/functions.html#enumerate)

```python
for i, declared_license in enumerate(declared_licenses):
    if declared_license.startswith("Version"):
```

File: src/graph_populator.py [hunk 4]
```diff
@@ -106,7 +106,16 @@ def construct_version_query(cls, input_json):
...
                     # string with comma separated license names
-                    declared_licenses = details[0]['declared_license'].split(',')
+                    # declared_licenses = details[0]['declared_license'].split(',')
+                    declared_licenses = details[0].get('declared_license').split(', ')
```
Comment (jpopelka) [inline src/graph_populator.py]: Not sure what's the intention for adding the `get()`, but I think it actually makes things worse.

Previously if there was no `'declared_license'` in `details[0]` you would get `KeyError: 'declared_license'`.
But with the suggested change you would in the same case get `AttributeError: 'NoneType' object has no attribute 'split'`, which hides the problem into more cryptic message.

File: src/graph_populator.py (hunk 1 above)
Comment (sunilk747) [reply src/graph_populator.py]: yeah. we can use the first approach that you have suggested.

Comment (sunilk747) [reply src/graph_populator.py]: @jpopelka Why we are checking for the same case in `elif`, i think the `if` is doing the job. 
https://github.com/sunilk747/fabric8-analytics-data-model/blob/49fd9a6ff7509f16e93042624b1dc5ec6fba3b0b/src/graph_populator.py#L104

Comment (jpopelka) [reply src/graph_populator.py]: Oh, you're right, good catch.
I probably didn't realize that when simplifying the statement in d067147c49
Luckily the comments there say it all, so I think it should look like:

```python
                declared_licenses = []
                dc = details[0].get('declared_licenses')
                if dc and isinstance(dc, list):
                    # list of license names
                    declared_licenses = dc
                elif dc and isinstance(dc, str):
                    # string with comma separated license names
                    declared_licenses = dc.split(',')
```

File: src/graph_populator.py (hunk 3 above)
Comment (jpopelka) [reply src/graph_populator.py]: See https://github.com/fabric8-analytics/fabric8-analytics-data-model/pull/70#discussion_r164379371

File: src/graph_populator.py (hunk 1 above)
Comment (sunilk747) [reply src/graph_populator.py]: Yep. Should I include these changes as well in my PR or you will do it?

Comment (jpopelka) [reply src/graph_populator.py]: In order to avoid conflicts it'd be better if you could do that (perhaps in separate commit).

File: src/graph_populator.py [hunk 5]
```diff
@@ -101,20 +101,20 @@ def construct_version_query(cls, input_json):
...
+                if dl and isinstance(dl, list):
                     # list of license names
-                    declared_licenses = details[0]['declared_licenses']
-                elif details[0].get('declared_license'):
+                    declared_licenses = dl
+                elif dl and isinstance(dl, str):
                     # string with comma separated license names
-                    # declared_licenses = details[0]['declared_license'].split(',')
-                    declared_licenses = details[0].get('declared_license').split(', ')
+                    declared_licenses = dl.replace(", ", ",").lower().split(",")
```
Comment (jpopelka) [inline src/graph_populator.py]: In that case you might perhaps want to also `replace(" ,", ",")` ?
There's a [strip()](https://docs.python.org/3/library/stdtypes.html#str.strip) method, which you can use to remove the white-spaces surrounding (not inside) the licenses, like this:
```python
declared_licenses = [l.strip() for l in dl.lower().split(",")]
```

File: src/graph_populator.py [hunk 6]
```diff
@@ -101,20 +101,20 @@ def construct_version_query(cls, input_json):
             details = input_json.get('analyses').get('metadata', {}).get('details', [])
             if details and details[0]:
                 declared_licenses = []
-                if details[0].get('declared_licenses'):
+                dl = details[0].get('declared_licenses')
+                if dl and isinstance(dl, list):
                     # list of license names
-                    declared_licenses = details[0]['declared_licenses']
-                elif details[0].get('declared_license'):
+                    declared_licenses = dl
```
Comment (jpopelka) [inline src/graph_populator.py]: In the `elif` branch you use `lower()`, so you probably want to use it also here ?
```python
declared_licenses = [l.lower() for l in dl]
```

File: src/graph_populator.py [hunk 7]
```diff
@@ -101,20 +101,20 @@ def construct_version_query(cls, input_json):
...
-                    for i in xrange(0, len(declared_licenses)):
-                        if declared_licenses[i].startswith("Version"):
+                    for declared_license in declared_licenses:
+                        if declared_license.startswith("version"):
                             final_declared_licenses[-1] = final_declared_licenses[-1] + ", "\
-                                                          + declared_licenses[i]
+                                                          + declared_license
                         else:
-                            final_declared_licenses.append(declared_licenses[i])
+                            final_declared_licenses.append(declared_licenses)
```
Comment (jpopelka) [inline src/graph_populator.py]: `declared_licenses` -> `declared_license`

File: src/graph_populator.py [hunk 8]
```diff
@@ -107,14 +107,15 @@ def construct_version_query(cls, input_json):
                     declared_licenses = dl
                 elif dl and isinstance(dl, str):
                     # string with comma separated license names
-                    declared_licenses = dl.replace(", ", ",").lower().split(",")
+                    declared_licenses = dl.replace(", ", ",").split(",")
                     final_declared_licenses = list()
                     for declared_license in declared_licenses:
-                        if declared_license.startswith("version"):
+                        if declared_license.startswith("version") or declared_license.\
+                                startswith("Version"):
```
Comment (jpopelka) [inline src/graph_populator.py]: prefix can also be a tuple of prefixes to look for, see https://docs.python.org/3/library/stdtypes.html#str.startswith

```python
if declared_license.startswith(("version", "Version")):
```

**Comment by jpopelka** [comment]
There's a conflict due to merged #72 
Please rebase.

File: src/graph_populator.py (hunk 8 above)
Comment (sunilk747) [reply src/graph_populator.py]: I think just converting the entire string to lowercase is a better approach no? ```python if declared_license.lower().startswith('version'): ```, looks more cleaner. What do you think?

**Comment by tuxdna** [comment]
@sunilk747  Can you incorporate `\n` checks in `declared_licenses` entries too? My PR https://github.com/fabric8-analytics/fabric8-analytics-data-model/pull/72 fixed it only for `declared_license`.

I suspect that `declared_licenses` entries could be containing newlines too as I see still see some errors in logs.

WDYT?

**Comment by sunilk747** [comment]
@tuxdna How `declared_licenses` and `declared_license ` are different here. I am a little confused. Please help me in understand.

**Comment by tuxdna** [comment]
@sunilk747 These are two different attributes as shown below:

https://github.com/fabric8-analytics/fabric8-analytics-data-model/blob/987ac37782aa3c2215640ce9a89189a422a56580/src/graph_populator.py#L104-L110

**Comment by jpopelka** [comment]
Aha, so it's `'declared_licenses'` vs. `'declared_license'`, thanks Saleem, you're the only one who's spotted that difference 👍 
In that case we can scratch my https://github.com/fabric8-analytics/fabric8-analytics-data-model/pull/70#discussion_r164379371.

File: src/graph_populator.py [hunk 9]
```diff
@@ -104,7 +116,7 @@ def construct_version_query(cls, input_json):
                 if details[0].get('declared_licenses'):
                     # list of license names
                     declared_licenses = details[0]['declared_licenses']
-                elif details[0].get('declared_license'):
+                elif details[0].get('declared_licenses'):
```
Comment (jpopelka) [inline src/graph_populator.py]: I'm a bit confused now too. Why this change ?

File: src/graph_populator.py [hunk 10]
```diff
@@ -134,6 +146,7 @@ def construct_version_query(cls, input_json):
                     else:  # default behavior
                         # split by comma
                         declared_licenses = [x.strip() for x in declared_str.split(",")]
+                    declared_licenses = cls.correct_license_splitting(declared_licenses)
```
Comment (jpopelka) [inline src/graph_populator.py]: I think you need to move this line 4 spaces left if you agree with [Saleem's suggestion](https://github.com/fabric8-analytics/fabric8-analytics-data-model/pull/70#issuecomment-361905406).

**Comment by tuxdna** [comment]
@sunilk747 Can you point me to the location list of licenses to be used for license analysis ?

**Comment by sunilk747** [comment]
@tuxdna https://github.com/fabric8-analytics/fabric8-analytics-license-analysis/blob/master/tests/synonyms/license_synonyms.json

**Comment by tuxdna** [comment]
@sunilk747 Thanks for the pointer. Can you also help me with the location(s) where this synonym mapping is used in license analysis ?

Thanks in advance!

**Comment by sunilk747** [comment]
@tuxdna https://github.com/fabric8-analytics/fabric8-analytics-license-analysis/blob/master/src/license_analysis.py#L59
If it looks good to you, merge it.

File: src/graph_populator.py [hunk 11]
```diff
@@ -11,6 +11,18 @@
 class GraphPopulator(object):
     """Class containing classmethods used to construct queries to the graph database."""
 
+    @classmethod
+    def correct_license_splitting(cls, license_list):
+        """Correct the incorrect splitting of licenses."""
+        final_declared_licenses = list()
+        for dl in license_list:
+            if dl.startswith(("version", "Version")):
```
Comment (tuxdna) [inline src/graph_populator.py]: Request you to add a transformation step that replaces all newlines with a space. 

```
" ".join(dl.split("\n"))
```

This will ensure that our queries do not break due to newlines. Otherwise looks good to me overall.

**Comment by tuxdna** [comment]
@sunilk747 PTAL my review comment.

File: src/graph_populator.py [hunk 12]
```diff
@@ -11,6 +11,18 @@
...
+                dl = " ".join([li.strip() for li in dl.split("\n")])
+                final_declared_licenses[-1] = final_declared_licenses[-1] + ", " + dl
+            else:
+                final_declared_licenses.append(dl)
```
Comment (tuxdna) [inline src/graph_populator.py]: Handle newlines in such a way that both code paths of `if` as well as `else` ensure newlines are not skipped.

Comment (sunilk747) [reply src/graph_populator.py]: done. PTAL :)

**Comment by sunilk747** [comment]
Thanks :) @tuxdna
